pydoctor/astbuilder.py

```diff
@@ -882,15 +899,20 @@ def add_arg(name: str, kind: Any, default: Optional[ast.expr]) -> None:
         if kwarg is not None:
             add_arg(kwarg.arg, Parameter.VAR_KEYWORD, None)
 
+        return_type = annotations.get('return')
+        return_annotation = Parameter.empty if return_type is None or is_none_literal(return_type) else _ValueFormatter(return_type, ctx=func)
         try:
-            signature = Signature(parameters)
+            signature = Signature(parameters, return_annotation=return_annotation)
         except ValueError as ex:
             func.report(f'{func.fullName()} has invalid parameters: {ex}')
             signature = Signature()
 
         func.signature = signature
-        func.annotations = self._annotations_from_function(node)
-    
+        func.annotations = annotations
+
+        if is_overload_func:
+            func.overloads.append(self.system.FunctionOverload(signature=func.signature))
```

Comment:
I think this implementation is OK, but could be more robust. 

The flow is just a bit weird because we're starting by creating the overloads (which makes sens because they are visited first) and incrementally append them to the `overloads` list of a `Function` object that is already created at the time we hit the last `FunctionDef` that contains the logic code of the function.

But what happens if a `@overload` decorator is encountered after the main declaration? I suppose the last signature overrides what what stored in `Fucntion.signature`. So maybe `Fucntion.signature` should be set only if the function is not an overloaded function? 

My understanding is that All `@overload` definitions must come before the implementation, and multiple implementations are not allowed. So when hitting an overloads function and the `Function.signature` is not `None` is means the `@overload` appeared after the main function def and we should warn about it, saying it's ignored.

Also what happens if a `@overload` function has a docstring? I'de suggest we leave a warning warning saying that docstrings on overloads function are ignored.

Tell me what you think

Reply:
> But what happens if a @overload decorator is encountered after the main declaration?

This is the same thought I had when I saw how we were pulling from `parent.contents`

> So maybe Function.signature should be set only if the function is not an overloaded function? [...] warn about it, saying it's ignored.

Works for me

> Also what happens if a @overload function has a docstring? I'de suggest we leave a warning warning saying that docstrings on overloads function are ignored.

This also works for me and is the path of least resistance.

I will write tests for these cases too (I have seen other tests that capture warning output)

Reply:
Not setting signature unless confirmed non-overload. Updated to warn on overload after primary and on overload with docstring.